code/src/reducers/products.js

```diff
@@ -0,0 +1,31 @@
+import { createSlice } from '@reduxjs/toolkit'
+
+export const products = createSlice({
+  name: 'products',
+  initialState: {
+    item: {},
+    scanStarted: false,
+  },
+
+  reducers: {
+    setProduct: (state, action) => {
+      state.item = action.payload
+    },
+    restart: (state) => {
+      state.scanStarted = false
+    },
+    isScanned: (state) => {
+      state.scanStarted = true
+    },
+  }
+})
+
+export const fetchProduct = (code) => {
+  return (dispatch) => {
+    fetch(`https://world.openfoodfacts.org/api/v0/product/${code}.json`)
+      .then((data) => data.json())
+      .then((json) => {
+        dispatch(products.actions.setProduct(json)) 
+    });
```

Comment:
Regarding your question about catching errors, if the response from the API is something in the range of 400-500 (400-499 errors are related to bad requests, 500-599 are related to server issues), according to [the fetch documentation](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch) the promise will not be rejected. Instead, you can check `response.ok` and react accordingly.

So, for example, say your products state has an `errorMessage` property and an action to set it, you could handle errors from the API in your thunk and dispatch the action to set the error message if something went wrong:

```js
fetch(`https://world.openfoodfacts.org/api/v0/product/${code}.json`)
  .then((response) => {
    if (response.ok) {
      // Only return the new promise if the API returned a good response
      return response.json()
    } else {
      // Dispatch the error action and don't return, so the promise chain stops
      dispatch(products.actions.setErrorMessage('Oh no, something went wrong'))
    }
  })
  .then((json) => {
    dispatch(products.actions.setProduct(json)) 
  })
```